src/legacy/server/saved_objects/service/lib/repository.ts

```diff
@@ -132,16 +194,20 @@ export class SavedObjectsRepository {
    * @property {string} [options.namespace]
    * @returns {promise} -  {saved_objects: [[{ id, type, version, references, attributes, error: { message } }]}
    */
-  async bulkCreate(objects, options = {}) {
+  async bulkCreate<T extends SavedObjectAttributes = any>(
+    objects: Array<BulkCreateObject<T>>,
+    options: CreateOptions = {}
+  ): Promise<BulkResponse<T>> {
     const { namespace, overwrite = false } = options;
     const time = this._getCurrentTime();
-    const bulkCreateParams = [];
+    const bulkCreateParams: object[] = [];
 
     let requestIndexCounter = 0;
-    const expectedResults = objects.map(object => {
-      if (!this._isTypeAllowed(object.type)) {
+    const expectedResults: Array<Either<any, any>> = objects.map(object => {
```

Comment:
Ideally `bulkCreate` should return an Either type so that typescript will enforce API clients correctly handling the error condition not just specifying the shape of the data. But that would break the API signature, so I only used it internally.

Reply:
with `Either<any, any>` we are loosing type information for error and data.

side note: `saved_objects` returned from `bulkCreate` should be something like `SavedObject | SavedObjectError` as those states are mutually exclusive, instead of mixing them in a common type https://github.com/elastic/kibana/pull/36829/files#diff-1882db7fdf2f1849465cc84f22a4273bR114

Reply:
I started work on returning `SavedObject | SavedObjectError` but I would have to change all the consuming code so it looked like it would make the PR even bigger. I do think this is really important so I've added a task to https://github.com/elastic/kibana/issues/33587

> Create stronger type safety with discriminated union and type guards for methods like `bulkGet` which return results and errors in a single array.

As part of this I'll also address your comment of making the `isError...` functions type guards and probably introduce a `isSavedObjectError` type guard.

To strictly type the the Left and Right hand Either, I will need to refactor this function a bit and I  think that will be easier once I've made the above changes.